serial-multiplexer/src/main.rs

```diff
@@ -42,130 +54,139 @@ fn main() {
+                port_receiver: port_receiver,
+            });
+
+            receiver_processors.push(SerialConnectionReceiverProcessor {
+                id: f.1.id,
+                port_manager: serial_port_manager_ref,
+                write_to_main_bus: main_bus_sender.clone(),
+            });
+
+            senders.push(SerialConnectionSender {
+                id: f.1.id,
+                port_sender: port_sender,
+            });
+        });
+    } else {
+        serial_ports_raw.iter().for_each(|f| {
+            let entry = f.1;
 
-                symlink(name, link_path).unwrap();
+            let (port_sender, port_receiver) = std::sync::mpsc::channel::<DataBlock>();
 
-                (
-                    entry.id as u32,
-                    config::SerialEntry {
-                        name: f.0.clone(),
-                        id: entry.id,
-                        device: master,
-                    }
-                )
-            })
-            .collect();
+            let (mut master, slave) = TTYPort::pair().expect("Unable to create ptty pair");
+            master.set_timeout(Duration::MAX).unwrap();
+
+            let name = slave.name().unwrap();
+            unused.push(slave);
+
+            let mut link_path = std::env::home_dir().unwrap_or(PathBuf::from("/dev"));
+
+            link_path.push("vtty");
+            if !link_path.exists() {
+                create_dir(&link_path).unwrap();
+            }
+
+            link_path.push(f.0);
+            let _ = remove_file(&link_path);
+
+            symlink(name, link_path).unwrap();
+
+            let serial_port_manager = SerialPortManager::with_port(master);
+            let serial_port_manager_ref = Arc::new(Mutex::new(serial_port_manager));
+
+            sender_processors.push(SerialConnectionSenderProcessor {
+                id: entry.id,
+                port_manager: serial_port_manager_ref.clone(),
+                port_receiver: port_receiver,
+            });
 
-        communicate(&mut multiplexed_port, &mut serial_ports);
+            receiver_processors.push(SerialConnectionReceiverProcessor {
+                id: entry.id,
+                port_manager: serial_port_manager_ref,
+                write_to_main_bus: main_bus_sender.clone(),
+            });
+
+            senders.push(SerialConnectionSender {
+                id: entry.id,
+                port_sender,
+            });
+        });
     }
+
+    println!("Starting communication loop...");
+    communicate(
+        sender_processors,
+        receiver_processors,
+        senders,
+        main_bus_receiver,
+        &mut multiplexed_port,
+    );
 }
 
 fn communicate(
+    sender_processors: Vec<SerialConnectionSenderProcessor>,
+    receiver_processors: Vec<SerialConnectionReceiverProcessor>,
+    senders: Vec<SerialConnectionSender>,
+    main_bus_receiver: Receiver<DataBlock>,
     multiplexed_port: &mut TTYPort,
-    serial_ports: &mut HashMap<u32, SerialEntry>,
 ) {
```

Comment:
I'd inline this into main, having this in a separate function doesn't give us any advantage.